aspnet/mvc/controllers/filters.rst

```diff
@@ -1,8 +1,312 @@
+
+.. literalinclude:: filters/sample/src/FiltersSample/Controllers/HomeController.cs
+  :language: c#
+  :emphasize-lines: 1-2
+  :lines: 33-39
+  :dedent: 8
+
+If you have a simple filter that doesn't require any arguments, but which has constructor dependencies that need to be filled by DI, you can inherit from ``TypeFilterAttribute``, allowing you to use your own named attribute on classes and methods (instead of ``[TypeFilter(typeof(FilterType))]``). The following filter shows how this can be implemented:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/SampleLoggingActionFilterAttribute.cs
+  :language: c#
+  :emphasize-lines: 1, 3, 7
+  :lines: 8-38
+  :dedent: 4
+
+This filter can be applied to classes or methods using the ``[SampleLoggingActionFilter]`` syntax.
+
+.. note:: Some of the filters shown in these examples implement logging in order to demonstrate when, where, or how they are executing. You should avoid creating and using filters purely for logging purposes, since the :doc:`built-in framework logging features </fundamentals/logging>` should already provide what you need to see how MVC is working under the covers. If you're going to add logging to your filters, it should focus on business domain concerns or behavior specific to your filter, rather than MVC actions or other framework events.
+
+``IFilterFactory`` actually implements ``IFilter``, so anywhere in the filter pipeline an ``IFilterFactory`` can be used in place of an ``IFilter`` instance. When it comes time for a filter to run, it is cast to an ``IFilterFactory``. If that cast succeeds, then the ``CreateInstance`` method is called to create the actual ``IFilter`` that will be used. This provides a very flexible design, since the precise filter pipeline does not need to be set in stone when the application starts.
+
+You can implement ``IFilterFactory`` on your own attribute implementations as another approach to creating filters:
+
+.. literalinclude:: filters/sample/src/FiltersSample/Filters/AddHeaderWithFactoryAttribute.cs
+  :language: c#
+  :emphasize-lines: 1,3-6
+  :lines: 6-25
+  :dedent: 4
+
+@rynowak - can you clarify this next bit?
+If you're using ``IFilterFactory``, you can specify lifetime of the filter (how?). When your filter is the instance, your filter instance is cached, so avoid do anything stateful in the filter returned by ``CreateInstance``.
+ 
+Ordering
+^^^^^^^^
+Filters can be applied per-action method (via attribute) or via controller (via attribute), or in global filters collection. Scope also generally determines ordering. The filter closest to the action runs first; generally you get overriding behavior without having to explicitly set ordering. This is sometimes referred to as "russian doll" nesting, as each increase in scope is wrapped around the previous scope, like a `nesting doll <https://en.wikipedia.org/wiki/Matryoshka_doll>`_.
```

Comment:
Filters can be applied to action methods or controllers (via attribute) or added to the global filters collection.